authentication-tutorial.md

```diff
@@ -0,0 +1,1314 @@
+before.
+
+![](https://loopback.io/pages/en/lb4/imgs/api_explorer_authorize_button.png)
+
+But now, **all** the endpoints have the lock icon.
+
+![](https://loopback.io/pages/en/lb4/imgs/api_explorer_all_sections_lock_icons1.png)
+
+This means that you can set the JWT token once via the
+`Authorize button/dialog`, and the token will be available to all the endpoints
+your interact with.
+
+There are plans to allow contributions to the OpenAPI specification via an
+extensionPoint/extensions pattern (
+[Issue #3854](https://github.com/strongloop/loopback-next/issues/3854) );
+including having authentication strategies automatically contribute security
+scheme/requirement object information (
+[Issue #3669](https://github.com/strongloop/loopback-next/issues/3669) ).
+
+### Summary
+
+We've gone through the steps that were used to add JWT `authentication` to the
+`loopback4-example-shopping` application, and add
+`security scheme/requirement object` settings to its OpenAPI specification.
+
+The final `ShoppingApplication` class in
+[loopback4-example-shopping/packages/shopping/src/application.ts](https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/application.ts)
+should look like this:
+
+```ts
+import {BootMixin} from '@loopback/boot';
+import {ApplicationConfig, BindingKey} from '@loopback/core';
+import {RepositoryMixin} from '@loopback/repository';
+import {RestApplication} from '@loopback/rest';
+import {ServiceMixin} from '@loopback/service-proxy';
+import {MyAuthenticationSequence} from './sequence';
+import {
+  RestExplorerBindings,
+  RestExplorerComponent,
+} from '@loopback/rest-explorer';
+import {
+  TokenServiceBindings,
+  UserServiceBindings,
+  TokenServiceConstants,
+} from './keys';
+import {JWTService} from './services/jwt-service';
+import {MyUserService} from './services/user-service';
+
+import path from 'path';
+import {
+  AuthenticationComponent,
+  registerAuthenticationStrategy,
+} from '@loopback/authentication';
+import {PasswordHasherBindings} from './keys';
+import {BcryptHasher} from './services/hash.password.bcryptjs';
+import {JWTAuthenticationStrategy} from './authentication-strategies/jwt-strategy';
+import {SECURITY_SCHEME_SPEC} from './utils/security-spec';
+
+/**
+ * Information from package.json
+ */
+export interface PackageInfo {
+  name: string;
+  version: string;
+  description: string;
+}
+export const PackageKey = BindingKey.create<PackageInfo>('application.package');
+
+const pkg: PackageInfo = require('../package.json');
+
+export class ShoppingApplication extends BootMixin(
+  ServiceMixin(RepositoryMixin(RestApplication)),
+) {
+  constructor(options?: ApplicationConfig) {
+    super(options);
+
+    this.api({
+      openapi: '3.0.0',
+      info: {title: pkg.name, version: pkg.version},
+      paths: {},
+      components: {securitySchemes: SECURITY_SCHEME_SPEC},
+      servers: [{url: '/'}],
+    });
```

Comment:
You may be able to simplify this part.  (see https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/application.ts).  In built enhancer in place now. 👍